### ISSUE TITLE
Don't leave comments with whitespace after removing annotations

### ISSUE THREAD

**Issue by sventschui** [terser/terser#526]
Over at preact we ship minified code but want the code to be tree-shakeable. While implementing this we faced some issues with inheritance. To solve this same issue babel, etc. construct classes/prototypes inside an IIFE that is marked as `/*#__PURE__*/`. We now want to do the same (see https://github.com/preactjs/preact/pull/2156#issuecomment-559913580) but face the issue of terser removing all annotations here:
https://github.com/terser/terser/blob/6589965fe372124ce94410c9e0e7389732a3c8ea/lib/output.js#L658

I'd like to ask whether you'd be open to extend terser with a config option to preserve these annotation comments? I'd be happy to PR this change.

Another thing I asked myself is whether this might have side effects. When something is inlined it might be dangerous to keep the `__INLINE__` annotation. I'm not super deep into the magic of minification so you might be able to answer this question quicker/better than me.

**Bug report or Feature request?**

Feature request

**Version (complete output of `terser -V` or specific git commit)**

n/a

**Complete CLI command or `minify()` options used**

Note the new option `--keep-annotations` that might take `all`, a regex or a list of strings

`terser --comments /__PURE__/ --keep-annotations all` 

**`terser` input**

```
import { Component } from 'preact'
const Suspense = /*#__PURE__*/(function createSuspense() {
  function Suspense() {}
  Suspense.prototype = new Component()
  return Suspense;
}())
```

**`terser` output or error**

```
import { Component } from 'preact'
const Suspense = /**/(function createSuspense() {
  function Suspense() {}
  Suspense.prototype = new Component()
  return Suspense;
}())
```

**Expected result**
```
import { Component } from 'preact'
const Suspense = /*#__PURE__*/(function createSuspense() {
  function Suspense() {}
  Suspense.prototype = new Component()
  return Suspense;
}())
```

**Comment by fabiosantoscode** [comment]
That's pretty reasonable :) if you're willing to PR this, please do go ahead! Thanks in advance!

**Comment by kzc** [comment]
Pure annotations were removed upon output on purpose. Comments can shift in a number of `compress` optimizations and can potentially lead to invalid code with non-pure calls inadvertently marked as pure and dropped. This most commonly happens when a user incorrectly annotates a non-call expression (such as a function declaration) as pure. Retaining pure annotations would be 100% safe if `compress` is disabled. Otherwise to do it correctly would require a fair amount of effort.

**Comment by kzc** [comment]
Here's another example of the difficulty of retaining pure annotations correctly - even when the annotations are used appropriately on calls:

Given:
```
$ cat annotation.js 
function f(x) {
    console.log(x);
    return console.log.bind(console);
}
(function() {
    var c = /*@__PURE__*/f(1);  // can't be pure since it is used below
    c(2);
    /*@__PURE__*/ f(3);
    f(4);
})();
```
we'd expect the minified output to produce:
```
$ cat annotation.js | terser -c | node
1
2
4
```
Now let's misspell the pure annotation to see what happens:
```
$ cat annotation.js | sed 's/PURE/PURe/g' | terser -c -b comments=/PUR/
function f(x) {
    return console.log(x), console.log.bind(console);
}

/*@__PURe__*/ f(1)(2), 
/*@__PURe__*/ f(3), f(4);
```
And if we convert `PURe` back to `PURE` and minify that we get the incorrect result:
```
$ cat annotation.js | sed 's/PURE/PURe/g' | terser -c -b comments=/PUR/ | sed 's/PURe/PURE/g' | terser -c | node
4
```

because pure annotations are greedy and choose the longest subsequent call (or `new`) expression.

There are many other different ways retaining pure annotations can go wrong if done incorrectly.

**Comment by tmorehouse** [comment]
I am finding that `/*#__PURE__*/` annotations (`/*#__PURE__*/`) are being rendered as `/* */` when I specify a `--comments="/#__PURE__/"`  cmd line option.   No matter how I modify the RegExpr string, I get empty comments `/* */` in the final output

**Comment by sventschui** [comment]
As @kzc pointed out: This is intentional as the annotations might be in the wrong place after minification. They‘re open for a PR but the solution is not as easy as just preserving them as one needs to make sure the annotations are at the right place after minification

**Comment by tmorehouse** [comment]
@sventschui I just find it bizarre that it renders `/* */` instead of nothing or `/*#__PURE__*/`

**Comment by fabiosantoscode** [comment]
That's very interesting. We can easily check if the comments are just whitespace after removing the annotation and remove them. Let's do that.

**Comment by tmorehouse** [comment]
@fabiosantoscode But I don't think that solves my problem of leaving `/*#__PURE__*/` comments alone (I want to leave them in our `esm` bundle for tree shaking by users).

**Comment by fabiosantoscode** [comment]
Yeah, I think we don't want to do that. At first I thought why not, then I read @kzc's reasoning and further thought of the maintenance burden of trying to move code with these annotations and preserve the annotations. Plus, I've been thinking of a new way of annotating which is a lot more resilient. It involves an npm module with do-nothing functions that have special meaning to the compiler. This approach of using comments works great if the only compiler reading them is UglifyJS/Terser and they are removed before output. To achieve something that can be read by multiple compilers a more solid approach is required.

**Comment by tmorehouse** [comment]
We currently have `--comments "/^!/"` to preserve banner comments, which works fine with terser, but single line comments like `/*#__PURE__*/`, when we set `--comments "/__PURE__/"` get converted to `/* */` in the output.

**Comment by kzc** [comment]
I advised the individuals who implemented pure annotations in Rollup to achieve a high degree of compatibility with uglify and its forks. Rollup does not perform most of the optimizations that uglify-js and terser do, so it manages to avoid the problems associated with retaining pure annotations (as far as I know). Please take the time to read and understand the posts above detailing why pure annotations were intentionally removed by uglify-js and terser to avoid incorrect results.

There other issues that I haven't mentioned in this thread. Even if my example above https://github.com/terser/terser/issues/526#issuecomment-560140628 were to hypothetically emit the following correctly annotated output with added parentheses:

```
$ cat annotation2.js
function f(x) {
    return console.log(x), console.log.bind(console);
}
(/*@__PURE__*/ f(1))(2), 
/*@__PURE__*/ f(3), f(4);
```
terser would continue to correctly process it:
```
$ cat annotation2.js | terser -bc
function f(x) {
    return console.log(x), console.log.bind(console);
}

f(1)(2), f(4);
```
```
$ cat annotation2.js | terser -bc | node
1
2
4
```
*however* Babel with default settings would not retain the necessary parentheses due to this [known issue](https://github.com/babel/babel/issues/10306):
```
$ cat annotation2.js | babel
"use strict";

function f(x) {
    return console.log(x), console.log.bind(console);
}
/*@__PURE__*/f(1)(2),
/*@__PURE__*/f(3), f(4);
```
so that when this Babel output would be (re)run through terser it would produce this undesirable result:
```
$ cat annotation2.js | babel | terser -bc
"use strict";

function f(x) {
    return console.log(x), console.log.bind(console);
}

f(4);
```
```
$ cat annotation2.js | babel | terser -bc | node
4
```
This is why uglify and its forks are as conservative as they are and elide pure annotations upon output. Workarounds can be devised for all of these issues, but they are not as trivial as they may appear.

**Comment by jridgewell** [comment]
Closing this as fixed with #550

**Comment by Finesse** [comment]
I have exactly the requirement that [sventschui](https://github.com/sventschui) described. I understand the risk of leaving the "pure" comments, and I want to leave the comments despite the risk. Thank you for making this possible.

A note for people from Google: besides adding a `comments` option you need to add a `preserve_annotations` option:

```js
const terserConfig = {
  module: true,
  format: {
    comments: /^\s*[#@]__PURE__\s*$/,
    preserve_annotations: true,
  },
}
```
